auto-revert SIGSEGV

**Description**

Levant crashes every time I deploy a job with a failing check when it gets to the auto-revert state.

By the way, why does Levant even do auto-revert, since Nomad already returns to a stable working version on its own? Can we add a flag to disable Levant's auto-revert functionality?

```
2021-09-27T23:58:41Z |INFO| levant/auto_revert: job the-job has entered auto-revert state; launching auto-revert checker job_id=the-job
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x77098e]

goroutine 1 [running]:
github.com/hashicorp/levant/levant.(*levantDeployment).autoRevert(0xc0003fd040, 0xc00011e2f0, 0xc00011e2d0)
	/home/circleci/project/project/levant/auto_revert.go:25 +0x16e
github.com/hashicorp/levant/levant.(*levantDeployment).checkAutoRevert(0xc0003fd040, 0xc00011e2d0)
	/home/circleci/project/project/levant/auto_revert.go:71 +0x186
github.com/hashicorp/levant/levant.(*levantDeployment).deploy(0xc0003fd040, 0x1)
	/home/circleci/project/project/levant/deploy.go:195 +0x505
github.com/hashicorp/levant/levant.TriggerDeployment(0xc0000ccd20, 0x0, 0xc0001696f0)
	/home/circleci/project/project/levant/deploy.go:81 +0x7e
github.com/hashicorp/levant/command.(*DeployCommand).Run(0xc0000aedf8, 0xc0000c0020, 0x3, 0x3, 0xc0000ccb00)
	/home/circleci/project/project/command/deploy.go:197 +0x939
github.com/mitchellh/cli.(*CLI).Run(0xc0000edcc0, 0xc0000edcc0, 0x7, 0xc0000aed08)
	/go/pkg/mod/github.com/mitchellh/cli@v1.1.0/cli.go:260 +0x41a
main.RunCustom(0xc0000c0010, 0x4, 0x4, 0xc000196870, 0x406365)
	/home/circleci/project/project/main.go:49 +0x33e
main.Run(0xc0000c0010, 0x4, 0x4, 0xc000094058)
	/home/circleci/project/project/main.go:17 +0x56
main.main()
	/home/circleci/project/project/main.go:11 +0x65
```


**Relevant Nomad job specification file**
I don't think this is very relevant, it happens every time.

**Output of `levant version`:**

```
Levant v0.3.0
```

**Output of `consul version`:**

```
Consul v1.10.0
Revision 27de64da7
Protocol 2 spoken by default, understands 2 to 3 (agent will automatically use protocol >2 when speaking to compatible agents)
```

**Output of `nomad version`:**

```
Nomad v1.1.2 (60638a086ef9630e2a9ba1e237e8426192a44244)
```

**Additional environment details:**

**Debug log outputs from Levant:**

<!--
If the log fragment is particularly long, please paste it into a gist or
similar and paste the link here.
-->

We also have similar issue when enabling **auto_revert** with `hashicorp/levant:0.3.1` docker

```
2022-12-07T04:19:10Z |INFO| levant/auto_revert: job develop--vas-service has entered auto-revert state; launching auto-revert checker job_id=develop--job
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x713dc4]
goroutine 1 [running]:
github.com/hashicorp/levant/levant.(*levantDeployment).autoRevert(0xc0003433f0, 0xc00043f2b0, 0xc00043f290)
	/home/circleci/project/project/levant/auto_revert.go:25 +0x[124](https://gitlab.private.domain/deco/vas/vas-service/-/jobs/1831#L124)
github.com/hashicorp/levant/levant.(*levantDeployment).checkAutoRevert(0xc000177bb0, 0xc00043f290)
	/home/circleci/project/project/levant/auto_revert.go:71 +0x129
github.com/hashicorp/levant/levant.(*levantDeployment).deploy(0xc0003433f0)
	/home/circleci/project/project/levant/deploy.go:193 +0x377
github.com/hashicorp/levant/levant.TriggerDeployment(0x7ffc04bfbf70, 0x4)
	/home/circleci/project/project/levant/deploy.go:81 +0x45
github.com/hashicorp/levant/command.(*DeployCommand).Run(0xc00000ce70, {0xc00001e0a0, 0x2, 0x2})
	/home/circleci/project/project/command/deploy.go:197 +0x934
github.com/mitchellh/cli.(*CLI).Run(0xc00022e000)
	/go/pkg/mod/github.com/mitchellh/cli@v1.1.0/cli.go:260 +0x5f8
main.RunCustom({0xc00001e090, 0xc0000f9f70, 0x405d79}, 0xc0001e0960)
	/home/circleci/project/project/main.go:49 +0x26a
main.Run({0xc00001e090, 0x3, 0x3})
	/home/circleci/project/project/main.go:17 +0x45
main.main()
	/home/circleci/project/project/main.go:11 +0x50
```

Is this normal / expected behavior on doing auto-revert?

--- a/levant/auto_revert.go
+++ b/levant/auto_revert.go
@@ -3,40 +3,40 @@ package levant
 import (
 	"time"
 
+	"github.com/hashicorp/nomad/api"
 	nomad "github.com/hashicorp/nomad/api"
 	"github.com/rs/zerolog/log"
 )
 
-func (l *levantDeployment) autoRevert(jobID, depID *string) {
+func (l *levantDeployment) autoRevert(dep *nomad.Deployment) {
 
 	// Setup a loop in order to retry a race condition whereby Levant may query
 	// the latest deployment (auto-revert dep) before it has been started.
 	i := 0
 	for i := 0; i < 5; i++ {
-
-		dep, _, err := l.nomad.Jobs().LatestDeployment(*jobID, nil)
+		revertDep, _, err := l.nomad.Jobs().LatestDeployment(dep.JobID, &api.QueryOptions{Namespace: dep.Namespace})
 		if err != nil {
-			log.Error().Msgf("levant/auto_revert: unable to query latest deployment of job %s", *jobID)
+			log.Error().Msgf("levant/auto_revert: unable to query latest deployment of job %s", dep.JobID)
 			return
 		}
 
 		// Check whether we have got the original deployment ID as a return from
 		// Nomad, and if so, continue the loop to try again.
-		if dep.ID == *depID {
-			log.Debug().Msgf("levant/auto_revert: auto-revert deployment not triggered for job %s, rechecking", *jobID)
+		if revertDep.ID == dep.ID {
+			log.Debug().Msgf("levant/auto_revert: auto-revert deployment not triggered for job %s, rechecking", dep.JobID)
 			time.Sleep(1 * time.Second)
 			continue
 		}
 
-		log.Info().Msgf("levant/auto_revert: beginning deployment watcher for job %s", *jobID)
-		success := l.deploymentWatcher(dep.ID)
+		log.Info().Msgf("levant/auto_revert: beginning deployment watcher for job %s", dep.JobID)
+		success := l.deploymentWatcher(revertDep.ID)
 
 		if success {
-			log.Info().Msgf("levant/auto_revert: auto-revert of job %s was successful", *jobID)
+			log.Info().Msgf("levant/auto_revert: auto-revert of job %s was successful", dep.JobID)
 			break
 		} else {
-			log.Error().Msgf("levant/auto_revert: auto-revert of job %s failed; POTENTIAL OUTAGE SITUATION", *jobID)
-			l.checkFailedDeployment(&dep.ID)
+			log.Error().Msgf("levant/auto_revert: auto-revert of job %s failed; POTENTIAL OUTAGE SITUATION", dep.JobID)
+			l.checkFailedDeployment(&revertDep.ID)
 			break
 		}
 	}
@@ -44,7 +44,7 @@ func (l *levantDeployment) autoRevert(jobID, depID *string) {
 	// At this point we have not been able to get the latest deploymentID that
 	// is different from the original so we can't perform auto-revert checking.
 	if i == 5 {
-		log.Error().Msgf("levant/auto_revert: unable to check auto-revert of job %s", *jobID)
+		log.Error().Msgf("levant/auto_revert: unable to check auto-revert of job %s", dep.JobID)
 	}
 }
 
@@ -60,6 +60,7 @@ func (l *levantDeployment) checkAutoRevert(dep *nomad.Deployment) {
 	for _, v := range dep.TaskGroups {
 		if v.AutoRevert {
 			revert = true
+			break
 		}
 	}
 
@@ -68,7 +69,7 @@ func (l *levantDeployment) checkAutoRevert(dep *nomad.Deployment) {
 			dep.JobID)
 
 		// Run the levant autoRevert function.
-		l.autoRevert(&dep.JobID, &dep.ID)
+		l.autoRevert(dep)
 	} else {
 		log.Info().Msgf("levant/auto_revert: job %v is not in auto-revert; POTENTIAL OUTAGE SITUATION", dep.JobID)
 	}

--- a/test/acctest/acctest.go
+++ b/test/acctest/acctest.go
@@ -12,6 +12,9 @@ type TestCase struct {
 	// Steps are ran in order stopping on failure
 	Steps []TestStep
 
+	// SetupFunc is called at before the steps
+	SetupFunc TestStateFunc
+
 	// CleanupFunc is called at the end of the TestCase
 	CleanupFunc TestStateFunc
 }
@@ -42,8 +45,9 @@ type TestStateFunc func(*TestState) error
 
 // TestState is the configuration for the TestCase
 type TestState struct {
-	JobName string
-	Nomad   *nomad.Client
+	JobName   string
+	Namespace string
+	Nomad     *nomad.Client
 }
 
 // Test executes a single TestCase
@@ -60,6 +64,12 @@ func Test(t *testing.T, c TestCase) {
 	state := &TestState{
 		JobName: fmt.Sprintf("levant-%s", t.Name()),
 		Nomad:   nomad,
+	}
+
+	if c.SetupFunc != nil {
+		if err := c.SetupFunc(state); err != nil {
+			t.Errorf("setup failed: %s", err)
+		}
 	}
 
 	for i, step := range c.Steps {
@@ -102,7 +112,7 @@ func Test(t *testing.T, c TestCase) {
 
 // CleanupPurgeJob is a cleanup func to purge the TestCase job from Nomad
 func CleanupPurgeJob(s *TestState) error {
-	_, _, err := s.Nomad.Jobs().Deregister(s.JobName, true, nil)
+	_, _, err := s.Nomad.Jobs().Deregister(s.JobName, true, &nomad.WriteOptions{Namespace: s.Namespace})
 	return err
 }
 
@@ -110,7 +120,7 @@ func CleanupPurgeJob(s *TestState) error {
 // the TestCase job matches the desired status
 func CheckDeploymentStatus(status string) TestStateFunc {
 	return func(s *TestState) error {
-		deploy, _, err := s.Nomad.Jobs().LatestDeployment(s.JobName, nil)
+		deploy, _, err := s.Nomad.Jobs().LatestDeployment(s.JobName, &nomad.QueryOptions{Namespace: s.Namespace})
 		if err != nil {
 			return err
 		}

--- a/test/acctest/deploy.go
+++ b/test/acctest/deploy.go
@@ -31,6 +31,10 @@ func (c DeployTestStepRunner) Run(s *TestState) error {
 		return fmt.Errorf("error rendering template: %s", err)
 	}
 
+	if job.Namespace != nil {
+		s.Namespace = *job.Namespace
+	}
+
 	cfg := &levant.DeployConfig{
 		Deploy: &structs.DeployConfig{
 			Canary:     c.Canary,

--- a/test/deploy_test.go
+++ b/test/deploy_test.go
@@ -1,9 +1,11 @@
 package test
 
 import (
+	"fmt"
 	"testing"
 
 	"github.com/hashicorp/levant/test/acctest"
+	"github.com/hashicorp/nomad/api"
 )
 
 func TestDeploy_basic(t *testing.T) {
@@ -17,6 +19,35 @@ func TestDeploy_basic(t *testing.T) {
 			},
 		},
 		CleanupFunc: acctest.CleanupPurgeJob,
+	})
+}
+
+func TestDeploy_namespace(t *testing.T) {
+	acctest.Test(t, acctest.TestCase{
+		SetupFunc: func(s *acctest.TestState) error {
+			if _, err := s.Nomad.Namespaces().Register(&api.Namespace{Name: "test"}, nil); err != nil {
+				return fmt.Errorf("could not create test namespace: %w", err)
+			}
+			return nil
+		},
+		Steps: []acctest.TestStep{
+			{
+				Runner: acctest.DeployTestStepRunner{
+					FixtureName: "deploy_namespace.nomad",
+				},
+				Check: acctest.CheckDeploymentStatus("successful"),
+			},
+		},
+		CleanupFunc: func(s *acctest.TestState) error {
+			if err := acctest.CleanupPurgeJob(s); err != nil {
+				return err
+			}
+
+			if _, err := s.Nomad.Namespaces().Delete("test", nil); err != nil {
+				return fmt.Errorf("could not delete namespace test: %w", err)
+			}
+			return nil
+		},
 	})
 }
 

--- a/test/fixtures/deploy_namespace.nomad
+++ b/test/fixtures/deploy_namespace.nomad
@@ -1,0 +1,43 @@
+# tests a healthy deployment
+
+job "[[.job_name]]" {
+  datacenters = ["dc1"]
+  type        = "service"
+  namespace   = "test"
+  update {
+    max_parallel     = 1
+    min_healthy_time = "10s"
+    healthy_deadline = "1m"
+    auto_revert      = true
+  }
+
+  group "test" {
+    count = 1
+    restart {
+      attempts = 10
+      interval = "5m"
+      delay    = "25s"
+      mode     = "delay"
+    }
+    ephemeral_disk {
+      size = 300
+    }
+    task "alpine" {
+      driver = "docker"
+      config {
+        image   = "alpine"
+        command = "tail"
+        args = [
+          "-f", "/dev/null"
+        ]
+      }
+      resources {
+        cpu    = 100
+        memory = 128
+        network {
+          mbits = 10
+        }
+      }
+    }
+  }
+}

